semgrep-core/src/engine/Match_patterns.ml

```diff
@@ -270,7 +271,25 @@ let check2 ~hook range_filter config rules equivs (file, lang, ast) =
              Common.push (pattern, strs, rule, cache) rules
            in
            match any with
-           | E pattern -> push_with_annotation any pattern expr_rules
+           | E pattern -> (
+               push_with_annotation any pattern expr_rules;
+               match pattern.e with
+               | Assign (_, tok, _) when config.Config.vardef_assign ->
+                   (* vardef-assign: let `$X = $E` match e.g. `var x = 1;`
+                    *
+                    * old: We used to have this equivalence built into
+                    * Visitor_AST. This was error prone because visitors
+                    * typically do side-effectful things and VarDefs were
+                    * visited twice (as a VarDef and as an Assign).
+                    *
+                    * If we just add an ExprStmt-pattern to stmt_rules,
+                    * the "vardef ==> assign" equivalence in
+                    * Generic_vs_generic.m_stmt will take care of the rest.
+                    * This will lead to some duplicate matches but that
+                    * should be OK (we already deal with duplicates). *)
+                   let s_pattern = s (ExprStmt (pattern, tok)) in
+                   push_with_annotation any s_pattern stmt_rules
```

Comment:
hmm but this will work only if the pattern is a single assign.
What if we have a more complex pattern like
```
foo()
...
x  = 1
```

Will this match code like
```
foo()
var x = 1
```
?

Reply:
Oh I got it, it's because we already do that something in m_stmts. It was written in your comment actually.
All good!
I don't even understand why we needed this visitor trick in the first place anymore ,,,

Reply:
I was thinking that this could indeed not work for deep expression patterns, e.g. matching `<... x = 1 ...>` against `var x = 1`, but it seems that that did not work before either.

Reply:
Oh I was making a silly mistake when testing this, yeah `<... x = 1 ...>` used to work and now it doesn't.

Reply:
Hmmm, I can't think of any other case than `<... $X = $E ...>` now but there could be others. Maybe we can just configure whether we want this equivalence or not when we create the visitor.

Reply:
if CI is passing, maybe we should just go for it even though we know it's more limited than before.
The cleaniness of Visitor_AST.ml seems more important to me, especially if there are no semgrep-rules
that fail with the change.
 (not sure why CI has failures now. Your first PR didn't have any).

Reply:
Oh the failure it's because I added a new test case and I used tabs instead of spaces in the YAML file.

Reply:
I updated this PR and took care of `<... $X = $E ...>`. f the pattern occurs in a place where we expect a statement then it's already going to be parsed as a `ExprStmt`, and if it occurs in a place where we expect an expression it should work unless that expression is a `Lambda` or something that contains statements.

I start to like more the alternative of adding a `bool` parameter to `mk_visitor`... I did the work here: https://github.com/returntocorp/semgrep/commit/daca5cceb7929e8e3a932e7219a7fafaf5672132. Anyways `Visitor_AST` still needs to deal with e.g. `Partial*`, so by taking out `vardef-assign` we are not simplifying it that much.